File: day-04-arrays-and-lists/python/concept04.py
"""
============================================================================
DAY 04 (Python) — Lists & Two Pointers
============================================================================

READ ME FIRST:
    This file is a lesson. Read it top to bottom, then run it:
        python3 concept04.py
    Then RETYPE the key snippets from memory in exercise04.py.

TODAY'S BIG IDEAS:
    1. Python lists ARE dynamic arrays — append is amortized O(1).
    2. In-place vs copying: know which methods mutate and which return new.
    3. Two pointers translate 1:1 from Java — same template, cleaner syntax.
    4. sort() mutates in place; sorted() returns new. Interviewers notice.
============================================================================
"""


def list_fundamentals():
    # ------------------------------------------------------------------
    # 1. LIST FUNDAMENTALS (vs Java arrays)
    # ------------------------------------------------------------------
    print("=== 1. List fundamentals ===")

    # Dynamic — no fixed size, no separate ArrayList. Heterogeneous too:
    items = [1, "two", 3.0]
    print("heterogeneous:", items)

    nums = [10, 20, 30]
    nums.append(40)            # amortized O(1) — Python over-allocates
    nums.insert(0, 5)          # O(n) — everything shifts (same as Java!)
    nums.pop()                 # remove from END: O(1)
    nums.pop(0)                # remove from FRONT: O(n)
    print("after ops:", nums)

    # Need fast front operations? collections.deque — O(1) both ends.
    # (Name-drop this in interviews; deep dive on Day 10 - Stacks & Queues.)

    # Negative indexing (Python-only superpower):
    print("last element:", nums[-1], "| first:", nums[0])

    # len() not .length / .size()
    print("length:", len(nums))


def in_place_vs_copy():
    # ------------------------------------------------------------------
    # 2. IN-PLACE vs COPY — the methods interviewers quiz you on
    # ------------------------------------------------------------------
    print("\n=== 2. In-place vs copy ===")

    a = [3, 1, 2]

    b = sorted(a)              # NEW list, a untouched
    print("sorted(a):", b, "| a still:", a)

    a.sort()                   # MUTATES a, returns None (!)
    print("a.sort(): ", a)

    # TRAP: writing  x = a.sort()  leaves x = None. Same for a.reverse().
    c = [1, 2, 3]
    c.reverse()                # in place
    print("c.reverse():", c)
    print("c[::-1] creates a NEW reversed list:", c[::-1], "| c unchanged:", c)

    # list.remove(x) removes the first VALUE x; del list[i] removes INDEX i.
    d = [1, 2, 3, 2]
    d.remove(2)                # removes first 2
    print("after remove(2):", d)

    # Copying (Day 01 callback — assignment never copies!):
    e = d                      # alias, NOT a copy
    f = d.copy()               # or d[:] or list(d)
    e.append(99)
    print("d after e.append:", d, "| f (real copy):", f)


def two_pointers():
    # ------------------------------------------------------------------
    # 3. TWO POINTERS — same template as Java
    # ------------------------------------------------------------------
    print("\n=== 3. Two pointers ===")

    def reverse_in_place(arr):
        left, right = 0, len(arr) - 1
        while left < right:
            arr[left], arr[right] = arr[right], arr[left]   # tuple swap — no tmp!
            left += 1
            right -= 1

    data = [1, 2, 3, 4, 5]
    reverse_in_place(data)
    print("reversed in place:", data)

    def two_sum_sorted(nums, target):
        """Opposite-direction pointers on a SORTED list -> indices or None."""
        left, right = 0, len(nums) - 1
        while left < right:
            s = nums[left] + nums[right]
            if s == target:
                return (left, right)
            if s < target:
                left += 1      # need bigger sum
            else:
                right -= 1     # need smaller sum
        return None

    print("pair summing to 9:", two_sum_sorted([1, 2, 4, 7, 11, 15], 9))

    def move_zeroes(nums):
        """Slow/fast flavor — same as Java version."""
        slow = 0
        for fast in range(len(nums)):
            if nums[fast] != 0:
                nums[slow], nums[fast] = nums[fast], nums[slow]
                slow += 1

    z = [0, 1, 0, 3, 12]
    move_zeroes(z)
    print("move_zeroes:", z)

    # Pythonic alternative — know both, be ready to defend the tradeoff:
    #   non_zero = [x for x in z if x != 0] + [0] * z.count(0)
    #   Cleaner, but O(n) extra space. Two pointers = O(1) space.


def pythonic_tools():
    # ------------------------------------------------------------------
    # 4. PYTHONIC TOOLS you'd use in real code (mention, don't always use)
    # ------------------------------------------------------------------
    print("\n=== 4. Pythonic tools ===")

    nums = [120, 85, 0, 200, 95, 0, 150]

    print("max/min/sum:", max(nums), min(nums), sum(nums))
    print("non-zero days:", [x for x in nums if x != 0])          # comprehension
    print("running max:", [max(nums[:i+1]) for i in range(3)], "... (O(n^2) — don't!)")

    # enumerate — index + value together (Day 02 callback):
    for i, v in enumerate(nums[:3]):
        print(f"  day {i}: {v}")

    # zip — parallel iteration:
    days = ["Mon", "Tue", "Wed"]
    for d, v in zip(days, nums):
        print(f"  {d}: {v}")


def interview_questions():
    print("\n=== Interview quick-fire (answer aloud!) ===")
    print("  Q: sort() vs sorted()?           A: mutates/None vs returns new list")
    print("  Q: append cost?                  A: amortized O(1) — over-allocation")
    print("  Q: Fast removal from front?      A: collections.deque — O(1)")
    print("  Q: x = mylist.sort() — bug?      A: yes, x is None")
    print("  Q: Two-sum on sorted data?       A: two pointers O(n), not brute force")


if __name__ == "__main__":
    list_fundamentals()
    in_place_vs_copy()
    two_pointers()
    pythonic_tools()
    interview_questions()

File: day-04-arrays-and-lists/java/Concept04.java
/**
 * ============================================================================
 * DAY 04 (Java) — Arrays & Lists
 * ============================================================================
 *
 * READ ME FIRST:
 *   This file is a lesson. Read it top to bottom, then run it:
 *       javac Concept04.java && java Concept04
 *   Then RETYPE the key snippets from memory in Exercise04.java.
 *
 * TODAY'S BIG IDEAS:
 *   1. Java arrays are FIXED size — know Arrays.asList vs ArrayList.
 *   2. "In-place" means O(1) extra space — swap, don't copy.
 *   3. TWO POINTERS: the #1 array interview pattern. Memorize the template.
 *   4. Front insertion is O(n); appends are (amortized) O(1).
 * ============================================================================
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Concept04 {

    public static void main(String[] args) {
        arraysVsArrayList();
        inPlaceReversal();
        twoPointerTemplate();
        costModel();
        interviewQuestions();
    }

    // ------------------------------------------------------------------
    // 1. ARRAYS vs ArrayList
    // ------------------------------------------------------------------
    static void arraysVsArrayList() {
        System.out.println("=== 1. Arrays vs ArrayList ===");

        // Fixed size, set at creation. Primitives allowed (int[]).
        int[] fixed = new int[3];              // {0, 0, 0} — zero-initialized!
        fixed[0] = 10;
        // fixed[3] = 99;                      // ArrayIndexOutOfBoundsException
        System.out.println("  fixed:    " + Arrays.toString(fixed) + " (length " + fixed.length + ")");

        // Dynamic, grows/shrinks. Objects only (Integer, not int — boxing!).
        List<Integer> dynamic = new ArrayList<>(List.of(10, 20, 30));
        dynamic.add(40);
        dynamic.remove(0);                     // shifts everything left — O(n)
        System.out.println("  dynamic:  " + dynamic + " (size " + dynamic.size() + ")");

        // GOTCHA interviews love: Arrays.asList returns a FIXED-SIZE list!
        List<Integer> trap = Arrays.asList(1, 2, 3);
        trap.set(0, 99);                       // OK — set works
        // trap.add(4);                        // UnsupportedOperationException!
        System.out.println("  Arrays.asList is fixed-size — add() throws. Wrap it:");
        System.out.println("  new ArrayList<>(Arrays.asList(1,2,3)) is the safe idiom");

        // Conversions (you'll type these a lot):
        int[] arr = {1, 2, 3};
        Integer[] boxed = Arrays.stream(arr).boxed().toArray(Integer[]::new);
        int[] back = Arrays.stream(boxed).mapToInt(Integer::intValue).toArray();
    }

    // ------------------------------------------------------------------
    // 2. IN-PLACE REVERSAL (the template everything builds on)
    // ------------------------------------------------------------------
    static void inPlaceReversal() {
        System.out.println("\n=== 2. In-place reversal (two pointers) ===");

        int[] a = {1, 2, 3, 4, 5};
        int left = 0, right = a.length - 1;
        while (left < right) {                 // stop when they MEET — not <=
            int tmp = a[left];
            a[left] = a[right];
            a[right] = tmp;
            left++;
            right--;
        }
        System.out.println("  reversed in place: " + Arrays.toString(a));

        // Why "in-place" matters: O(1) extra space vs O(n) for a copy.
        // Say the complexity out loud whenever you write one:
        //   time O(n) — each element swapped once
        //   space O(1) — only tmp + two indices
    }

    // ------------------------------------------------------------------
    // 3. THE TWO-POINTER TEMPLATE (and its two flavors)
    // ------------------------------------------------------------------
    static void twoPointerTemplate() {
        System.out.println("\n=== 3. Two-pointer flavors ===");

        // FLAVOR A: pointers move TOWARD each other (reverse, palindrome, pair-sum)
        //   left=0, right=n-1; while (left < right) { ...; left++; right--; }

        // FLAVOR B: slow/fast, same direction (dedupe in place, move zeroes)
        int[] nums = {0, 1, 0, 3, 12};
        moveZeroes(nums);
        System.out.println("  moveZeroes:  " + Arrays.toString(nums));   // [1,3,12,0,0]

        // Classic pair-sum on a SORTED array — O(n) instead of brute O(n^2):
        int[] sorted = {1, 2, 4, 7, 11, 15};
        System.out.println("  pair summing to 9: " + Arrays.toString(twoSumSorted(sorted, 9)));
    }

    // Slow/fast: `slow` marks where the next non-zero goes.
    static void moveZeroes(int[] nums) {
        int slow = 0;
        for (int fast = 0; fast < nums.length; fast++) {
            if (nums[fast] != 0) {
                int tmp = nums[slow];
                nums[slow] = nums[fast];
                nums[fast] = tmp;
                slow++;
            }
        }
    }

    // Opposite-direction: shrink the window by logic, not by scanning.
    static int[] twoSumSorted(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int sum = nums[left] + nums[right];
            if (sum == target) return new int[]{left, right};
            if (sum < target) left++;        // need a bigger sum
            else right--;                    // need a smaller sum
        }
        return new int[]{-1, -1};
    }

    // ------------------------------------------------------------------
    // 4. COST MODEL (say these out loud until automatic)
    // ------------------------------------------------------------------
    static void costModel() {
        System.out.println("\n=== 4. Cost model ===");
        System.out.println("  access by index        O(1)   — contiguous memory");
        System.out.println("  append at end          O(1)*  — *amortized for ArrayList");
        System.out.println("  insert/remove at front O(n)   — everything shifts");
        System.out.println("  search unsorted        O(n)   — linear scan");
        System.out.println("  search sorted          O(log n) — binary search (Day 25!)");
        System.out.println("  sort                   O(n log n) — Arrays.sort / Collections.sort");

        // Sorting both ways:
        int[] a = {5, 2, 8, 1};
        Arrays.sort(a);                                   // in place, ascending
        System.out.println("  sorted ascending:  " + Arrays.toString(a));

        Integer[] b = {5, 2, 8, 1};
        Arrays.sort(b, (x, y) -> y - x);                  // descending needs Comparator
        System.out.println("  sorted descending: " + Arrays.toString(b));
        // NOTE: primitives (int[]) can't take a Comparator — another boxing gotcha.
    }

    // ------------------------------------------------------------------
    // SAY THESE OUT LOUD
    // ------------------------------------------------------------------
    static void interviewQuestions() {
        System.out.println("\n=== Interview quick-fire (answer aloud!) ===");
        System.out.println("  Q: Array full — now what?            A: ArrayList (or allocate new + copy)");
        System.out.println("  Q: Reverse in place, extra space?    A: O(1) — two pointers + swap");
        System.out.println("  Q: When do two pointers stop?        A: when they meet: left < right");
        System.out.println("  Q: Insert at front cost?             A: O(n) — all elements shift");
        System.out.println("  Q: Pair-sum on sorted array?         A: two pointers O(n), not brute O(n^2)");
    }
}
